processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java

```diff
@@ -0,0 +1,536 @@
+            return comparator.compare(
+                curHashTableBuffer,
+                curHashTableBuffer,
+                lhs + HASH_SIZE,
+                rhs + HASH_SIZE
+            );
+          }
+        }
+    );
+
+    return new Iterator<Grouper.Entry<KeyType>>()
+    {
+      int curr = 0;
+
+      @Override
+      public boolean hasNext()
+      {
+        return curr < size;
+      }
+
+      @Override
+      public Grouper.Entry<KeyType> next()
+      {
+        return bucketEntryForOffset(wrappedOffsets.get(curr++));
+      }
+
+      @Override
+      public void remove()
+      {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  private Iterator<Grouper.Entry<KeyType>> makeHeapIterator()
+  {
+    final int initialHeapSize = offsetHeap.getHeapSize();
+    return new Iterator<Grouper.Entry<KeyType>>()
+    {
+      int curr = 0;
+
+      @Override
+      public boolean hasNext()
+      {
+        return curr < initialHeapSize;
+      }
+
+      @Override
+      public Grouper.Entry<KeyType> next()
+      {
+        if (curr >= initialHeapSize) {
+          throw new NoSuchElementException();
+        }
+        final int offset = offsetHeap.removeMin();
+        final Grouper.Entry<KeyType> entry = bucketEntryForOffset(offset);
+        curr++;
+
+        return entry;
+      }
+
+      @Override
+      public void remove()
+      {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  private Comparator<Integer> makeHeapComparator()
+  {
+    return new Comparator<Integer>()
+    {
+      final BufferComparator bufferComparator = keySerde.bufferComparatorWithAggregators(
+          aggregatorFactories,
+          aggregatorOffsets
+      );
+      @Override
+      public int compare(Integer o1, Integer o2)
+      {
+        final ByteBuffer tableBuffer = hashTable.getTableBuffer();
+        return bufferComparator.compare(tableBuffer, tableBuffer, o1 + HASH_SIZE, o2 + HASH_SIZE);
+      }
+    };
+  }
+
+
+  public static void validateBufferCapacity(
+      int limit,
+      float maxLoadFactor,
+      ByteBuffer buffer,
+      int bucketSize
+  )
+  {
+    int numBucketsNeeded = (int) Math.ceil((limit + 1) / maxLoadFactor);
+    int targetTableArenaSize = numBucketsNeeded * bucketSize * 2;
+    int heapSize = (limit + 1) * (Ints.BYTES);
+    int requiredSize = targetTableArenaSize + heapSize;
+
+    if (buffer.capacity() < requiredSize) {
+      throw new IAE(
+          "Buffer capacity [%d] is too small for limit[%d] with load factor[%f], minimum bytes needed: [%d]",
+          buffer.capacity(),
+          limit,
+          maxLoadFactor,
+          requiredSize
+      );
+    }
+  }
+
+  private class AlternatingByteBufferHashTable extends ByteBufferHashTable
+  {
+    private ByteBuffer[] subHashTableBuffers;
+    private ByteBufferHashTable[] subHashTables;
+    private ByteBufferHashTable activeHashTable;
```

Comment:
Changed this to use two ByteBuffers directly without the sub HashTables, added a comment on there being two buffers and a description of the swapping